portal/index.php

```diff
@@ -35,7 +38,33 @@
     echo xlt('Patient Portal is turned off');
     exit;
 }
-
+$auth['portal_pwd'] = '';
+if (isset($_GET['woops'])) {
+    unset($_GET['woops']);
+    unset($_SESSION['password_update']);
+}
+if (isset($_GET['forward'])) {
+    $auth = sqlQueryNoLog("Select * From patient_access_onsite Where portal_login_username = ?", array($_GET['forward']));
```

Comment:
Doesn't this mean that the portal_login_username gets cleared whenever users has their credentials reset. This seems odd. You are going to hate me, but why not store the short-term token in it's own sql column. Then portal_login_username can persist after credential reset (this also solves weird issues if portal_login_username is same as another user's portal_username in get_patient_info.php); then on upgrade would just make portal_login_username same as portal_username, which means no need to have the legacy support queries in get_patient_info.php.

Reply:
By design. I want login name cleared and have user recommit to credentials. I do get paranoid about usernames. This prevents anyone that may know a username from trying to guess at a reset password. For a patient that set their own password, a weak password that gets hacked is on them however, where an admin resets a portal password and it gets hacked then that is on us....
It may seem odd but this is open source and anyone looking at how we manage resets will know a username still exists and will also know new pass is fixed at given number of characters making life so much easier. Using the one time token will ensure no one will guess a username.

Agreed though that legacy login has to go away asap. One, I don't want portal account used in login username. Two, i'll rework login for login name and last, i'll force all patients to reset their passwords after this update by resetting flag in table on upgrade.

Reply:
What happens here if the request includes $_GET['forward'] and a valid patient login name in $_GET['forward'] (ie. and not a token)? Makes me wonder if it would separate out the token into a separate sql column. Then wouldn't have to be worried about this type of attack.

Reply:
It'd fail. One, it would confuse heck out of them with token expired and if try to resolve forward param length as login name, just kicks back to login and logs to core logs.

Wow, you really hate I use login name:) If it's confusing, it is meant to be. I get what a normal pattern for one time is usually and if OpenEMR was consistent throughout with passwords, then a separate table to manage dynamic password operations and history would be worthy.

Adding a new one time field to password table may seem to solve some security problems with login names but, too me, adding a new field just for a one time token opens up new and different set of concerns.